packages/backend/src/queryBuilder.ts

```diff
@@ -26,18 +28,37 @@ import {
     renderTableCalculationFilterRuleSql,
     SortField,
     SupportedDbtAdapter,
+    TimeFrames,
     UserAttributeValueMap,
+    validateTimeFrames,
     WarehouseClient,
 } from '@lightdash/common';
 import { hasUserAttribute } from './services/UserAttributesService/UserAttributeUtils';
 
-const getDimensionFromId = (dimId: FieldId, explore: Explore) => {
+const getDimensionFromId = (
+    dimId: FieldId,
+    explore: Explore,
+): CompiledDimension => {
     const dimensions = getDimensions(explore);
     const dimension = dimensions.find((d) => fieldId(d) === dimId);
-    if (dimension === undefined)
+
+    if (dimension === undefined) {
+        const { baseDimensionId, newTimeFrame } = getDateDimension(dimId);
+
+        if (baseDimensionId) {
+            const baseField = getDimensionFromId(baseDimensionId, explore);
+            if (baseField && newTimeFrame)
+                return {
+                    ...baseField,
+                    compiledSql: `DATE_TRUNC('${newTimeFrame}', ${baseField.compiledSql})`,
```

Comment:
Creates a missing time frame on the fly

Reply:
Would there be a case where the code inside the if statement (L46-55) could be run without the intent to Date Zoom? 
For example, a date dimension that can't be found, but we return a new one with the `missing time frame on the fly` but the user action is not to `date zoom` but instead compile a project or something?